FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+
+```python
+>>> xr.DataArray([1.0, 2.0], dims='x', coords={'x': midx})
+<xarray.DataArray (x: 2)>
+array([1., 2.])
+Coordinates:
+    x        (x) object ('a', 0) ('b', 1)
+```
+
+A possible, more explicit solution to reuse a `pandas.MultiIndex` in a DataArray/Dataset with levels exposed as coordinates is proposed in [Section 2.2.4](#224-implicit-coordinates).
+
+#### 2.2.2 Dataset/DataArray's `set_index` method
+
+New indexes may also be built from existing sets of coordinates or variables in a Dataset/DataArray using the `.set_index()` method.
+
+The [current signature](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.set_index.html#xarray.DataArray.set_index) of `.set_index()` is tailored to `pandas.MultiIndex` and tied to the concept of a dimension-index. It is therefore hardly reusable as-is in the context of flexible indexes proposed here.
+
+The new signature may look like one of these:
+
+- A. `.set_index(coords: CoordinateNames, index: Union[XarrayIndex, Type[XarrayIndex]], **index_kwargs)`: one index is set at a time, index construction options may be passed as keyword arguments
+- B. `.set_index(indexes: Mapping[CoordinateNames, Union[Type[XarrayIndex], Tuple[Type[XarrayIndex], Dict[str, Any]]]])`: multiple indexes may be set at a time from a mapping of coordinate or variable name(s) as keys and `XarrayIndex` subclasses (maybe with a dict of build options) as values. If variable names are given as keys of they will be promoted as coordinates
+
+Option A looks simple and elegant but significantly departs from the current signature. Option B is more consistent with the Dataset/DataArray constructor signature proposed in the previous section and would be easier to adopt in parallel with the current signature that we could still support through some depreciation cycle.
+
+The `append` parameter of the current `.set_index()` is specific to `pandas.MultiIndex`. With option B we could still support it, although we might want to either drop it or move it to the index construction options in the future.
+
+#### 2.2.3 Implicit default indexes
+
+In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.
+
+For which coordinates?
+
+- A. only 1D coordinates with a name matching their dimension name
+- B. all 1D coordinates
```

Comment:
:+1: to all 1D coordinates. I call `swap_dims` often just to make it easier to select based on different 1D coordinates. This would eliminate that. But could get expensive for datasets with lots of long non-dimension coordinates.

Also, automatically creating indexes could trigger eager loading of coordinates that would otherwise be lazy. For some datasets this would be catastrophic for memory usage (related to #1650).

Reply:
Once we have some kind of out-of-core index in Xarray (as suggested in #1650) I guess we could safely extend the implicit index creation to all 1D coordinates, but in the meantime it's probably wiser to limit that to dimension coordinates for the eager loading issues that you mention. Instead of calling `swap_dims`, you'll just need one explicit call to `set_index`.